[Lens] Add embeddable transforms

nickofthyme opened the pull request:
## Summary

**_Branched off of #229534_** See https://github.com/nickofthyme/kibana/pull/7 for simplified code diff.

**Lens transforms Phase 2** - Lens by-value transforms for dashboard and standalone embeddables.

Adds new Lens embeddable transform. This initially only transforms current Lens Serialized state attributes with old configurations to the latest Serialized state attributes. Applying all existing runtime migrations and pins version to latest Lens item.

Additionally, this implements the embeddable transforms on the embeddable renderer to handle standalone usages of the Lens embeddable, an extension of the work done in #227141.

Closes #229543

### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.

nickofthyme committed:
improve CM class to be easier to extend and override defaults

nickofthyme committed:
improve CM rpc types to satisfy generic type instead of be the generic type

nickofthyme committed:
fix spelling errors

nickofthyme committed:
move and add constants

nickofthyme committed:
update CM common schema

- make accessible from /common
- fix reference.name schema misalignment
- add `createdBy` and `updatedBy` to the SO schema
- use generics on common schema functions to pass type correctly
- add schema for search result

nickofthyme committed:
update storage context params

- The `getContentClientFactory` required request but never uses it.
- Add `request` to `StorageContext` to pass to CM storage handlers

nickofthyme committed:
cleanup `getMSearchClientFactory`

- This never checked if `mSearch` was available when creating the ctx.
- Check for `mSearch` before adding to ctx, log server warning if missing
- Add `logger` to `mSearch` factory

nickofthyme committed:
use api path constants

nickofthyme committed:
allow initial `v0` for `down` transforms

- We should allow transforms to run on unversioned items
- These version are not targets, they are only pulled up to the first version.
- allow using `kbn/object-versioning` in common.

nickofthyme committed:
refactor visualizations plugin clients

- Vis type alias client only needs basic methods (i.e. `get` and `update`).
- Create a slim `BasicVisualizationClient`
- The basic client can use `cm` or `http`.

nickofthyme committed:
add content management schema

nickofthyme committed:
add lens vis route schema

nickofthyme committed:
refactor types to use schema definitions

nickofthyme committed:
update imports, constants and types

nickofthyme committed:
update lens SO mappings to add new version property

nickofthyme committed:
implement all lens CM CRUD operations with latest CM schema

- isolates the Lens SO to CM

nickofthyme committed:
update lens CRUD apis

- update constants and use route schema
- update handling of API formats in and out of CM client

nickofthyme committed:
update CM service with v0 Item transforms

- This add CM transforms to convert existing lens items to v1
- Adds version number to lens SO attributes.
- Duplicates and tweaks runtime migrations for legend and color mappings.

nickofthyme committed:
add config builder stub

nickofthyme committed:
cleanup public constants and types

nickofthyme committed:
refactor lens public client

- Refactor lens client to use new CRUD api instead of CM
- Create basic lens client using CM to use new API types
- Convert lens type alias client to use new basic client
- Update calls to lens client dependency to `http`
- Update lens doc service to use new lens client
- Update lens attribute service to use updated doc service

nickofthyme committed:
more type cleanup

nickofthyme committed:
register lens embeddable transforms

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

nickofthyme committed:
fix build error

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/styled_components_mapping'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

nickofthyme committed:
add sync check for transform before fetching

nickofthyme committed:
Merge branch 'main' into lens-transforms-phase-1

nickofthyme committed:
chore: fix type check and linting errors

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/yarn_deduplicate'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

nickofthyme committed:
chore: fix more type errors

- remove unused mSearch helper
- update abstract storage class tests with stricter types

nickofthyme committed:
Merge branch 'main' into lens-transforms-phase-1

nickofthyme committed:
Merge branch 'lens-transforms-phase-1' into lens-transforms-phase-2

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/notice'

nickofthyme committed:
fix: import error

nickofthyme committed:
allow accessing props outside of schema

nickofthyme committed:
Merge branch 'main' into lens-transforms-phase-1

nickofthyme committed:
fix test and type errors

- update lens doc types to match previous
- update mixed null/undefined values
- update lens doc service tests

nickofthyme committed:
Merge branch 'main' into lens-transforms-phase-1

nickofthyme committed:
fix type and linting issues

nickofthyme committed:
update register type

nickofthyme committed:
fix more type issues

nickofthyme committed:
fix more type errors and version issues

nickofthyme committed:
remove erroneous lens attributes

nickofthyme committed:
fix another type issue

nickofthyme committed:
fix lens attribute versioning

nickofthyme committed:
fix bad imports

nickofthyme committed:
Merge branch 'main' into lens-transforms-phase-1

nickofthyme committed:
permit unknown SO versioning

- an existing SO import should not fail if no version is supplied
- save old lens SO as is and handle migration when retrieved and updates.

nickofthyme committed:
fix another build errors

nickofthyme committed:
Merge remote-tracking branch 'origin/lens-transforms-phase-1' into lens-transforms-phase-1

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

nickofthyme committed:
fix: permit old SO shape on create

- Create may still be called with the old state such as saving an old lens by-value dashboard panel.

nickofthyme committed:
fix: lens doc services should return doc from request

- the original attributes may have changed, or transformed. The request will always return the correct updated attributes.

nickofthyme committed:
Merge remote-tracking branch 'origin/lens-transforms-phase-1' into lens-transforms-phase-1

nickofthyme committed:
revert required version attribute changes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

nickofthyme committed:
chore: remove versions in cb files

nickofthyme committed:
temporary limit bump

nickofthyme committed:
make version optional

nickofthyme committed:
allow passing old SO to create route

nickofthyme committed:
chore: remove erroneous props

nickofthyme committed:
Merge remote-tracking branch 'origin/lens-transforms-phase-1' into lens-transforms-phase-1

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

nickofthyme committed:
chore: move CM files and schema from common/ to server/

- we don't need schemas in common only their types
- using schema in common pulls in joi to the bundle
- keep constants in common and import in server, public

nickofthyme committed:
Revert "temporary limit bump"

This reverts commit fe582c046881eb1dded6b94a62ffb48d2c080019.

nickofthyme committed:
update registered types

nickofthyme committed:
Merge remote-tracking branch 'origin/lens-transforms-phase-1' into lens-transforms-phase-1

nickofthyme committed:
fix: lens save error

nickofthyme committed:
Merge branch 'main' into lens-transforms-phase-1

nickofthyme committed:
Merge branch 'lens-transforms-phase-1' into lens-transforms-phase-2

nickofthyme commented:
Closed in favor of #235277
